lib/temporal/workflow/future.rb

```diff
@@ -52,14 +52,17 @@ def fail(exception)
         @failed = true
       end
 
+      # When the activity completes successfully, the block will be called with any result
       def done(&block)
-        # do nothing
-        return if failed?
+        add_callback do |result, _|
+          block.call(result) if ready?
+        end
+      end
 
-        if ready?
-          block.call(result)
-        else
-          callbacks << block
+      # When the activity fails, the block will be called with the exception
+      def failed(&block)
+        add_callback do |_, exception|
```

Comment:
@jeffschoner-stripe I wonder what are your thoughts on simplifying this by having 2 sets of callback array internally — one for success and one for failure cases? From what you're doing it doesn't look like there's a lot in common between them, which makes it a bit more confusing

Reply:
I'll give that a try. I've iterated from an earlier version of this that had just one method for registering a callback for both successes and failures. This is probably a suboptimal remnant of that earlier design.